restcomm/restcomm.commons/src/main/java/org/restcomm/connect/commons/util/DigestAuthentication.java

```diff
@@ -46,31 +46,53 @@ private static String A1(final String algorithm, final String user, final String
     }
 
     private static String A2(final String method, final String uri, String body, final String qop) {
+        return A2("MD5", method, uri, body, qop);
+    }
+
+    private static String A2(String algorithm, final String method, final String uri, String body, final String qop) {
         if (qop == null || qop.trim().length() == 0 || qop.trim().equalsIgnoreCase("auth")) {
             return method + ":" + uri;
         } else {
             if (body == null)
                 body = "";
-            return method + ":" + uri + ":" + H(body);
+            return method + ":" + uri + ":" + H(body, algorithm);
         }
     }
 
     public static String response(final String algorithm, final String user, final String realm, final String password,
-            final String nonce, final String nc, final String cnonce, final String method, final String uri, String body,
-            final String qop) {
-        validate(user, realm, password, nonce, method, uri);
-        final String a1 = A1(algorithm, user, realm, password, nonce, cnonce);
-        final String a2 = A2(method, uri, body, qop);
+    final String nonce, final String nc, final String cnonce, final String method, final String uri, String body,
+    final String qop) {
+        return response(algorithm, user, realm, password, "", nonce, nc, cnonce, method, uri, body, qop);
+    }
+
+    public static String response(final String algorithm, final String user, final String realm, final String password,
+            String password2, final String nonce, final String nc, final String cnonce, final String method, final String uri,
+            String body, final String qop) {
+        validate(user, realm, password, nonce, method, uri, algorithm);
+        String ha1;
+
+        if(!password2.isEmpty()){
+            ha1 = password2;
+        }else{
+            final String a1 = A1(algorithm, user, realm, password, nonce, cnonce);
+            ha1 = H(a1, algorithm);
+        }
+
+        final String a2 = A2(algorithm, method, uri, body, qop);
         if (cnonce != null && qop != null && nc != null && (qop.equalsIgnoreCase("auth") || qop.equalsIgnoreCase("auth-int"))) {
-            return KD(H(a1), nonce + ":" + nc + ":" + cnonce + ":" + qop + ":" + H(a2));
+            return KD(ha1, nonce + ":" + nc + ":" + cnonce + ":" + qop + ":" + H(a2, algorithm), algorithm);
         } else {
-            return KD(H(a1), nonce + ":" + H(a2));
+            return KD(ha1, nonce + ":" + H(a2, algorithm), algorithm);
         }
     }
 
     private static String H(final String data) {
+        return H(data, "MD5");
```

Comment:
@abdulazizali77 why do we hardcode to MD5 instead of using client-algorithm from the settings ?